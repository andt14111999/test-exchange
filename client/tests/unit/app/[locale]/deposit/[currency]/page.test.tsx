import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import DepositPage from "@/app/[locale]/deposit/[currency]/page";

// Mock next-intl
jest.mock("next-intl", () => ({
  useTranslations: () => (key: string) => key,
}));

// Mock next/navigation
const push = jest.fn();
jest.mock("next/navigation", () => ({
  useRouter: () => ({ push }),
  useParams: () => ({ currency: "vnd" }),
}));

// Mock toast
const toast = jest.fn();
jest.mock("@/components/ui/use-toast", () => ({
  useToast: () => ({ toast }),
}));

// Mock useWallet
jest.mock("@/hooks/use-wallet", () => ({
  useWallet: () => ({
    data: { fiat_accounts: [{ currency: "vnd", balance: 1000000 }] },
  }),
}));

// Mock useUserStore
jest.mock("@/lib/store/user-store", () => ({
  useUserStore: () => ({ user: { id: 1 } }),
}));

// Mock API
const getOffers = jest.fn();
jest.mock("@/lib/api/merchant", () => ({
  getOffers: () => getOffers(),
}));
const createTrade = jest.fn();
jest.mock("@/lib/api/trades", () => ({
  createTrade: (params: unknown) => createTrade(params),
}));

// Mock layout components
jest.mock("@/components/fiat-transaction-layout", () => ({
  FiatTransactionLayout: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="fiat-transaction-layout">{children}</div>
  ),
}));
jest.mock("@/components/protected-layout", () => ({
  ProtectedLayout: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="protected-layout">{children}</div>
  ),
}));

// Mock UI components
jest.mock("@/components/ui/button", () => ({
  Button: ({
    children,
    asChild,
    ...props
  }: { children: React.ReactNode; asChild?: boolean } & Record<
    string,
    unknown
  >) => {
    if (asChild) {
      return <button {...props}>{children}</button>;
    }
    return <button {...props}>{children}</button>;
  },
}));
jest.mock("@/components/ui/input", () => ({
  Input: (props: Record<string, unknown>) => <input {...props} />,
}));
jest.mock("@/components/ui/label", () => ({
  Label: (props: Record<string, unknown>) => <label {...props} />,
}));
jest.mock("@/components/ui/card", () => ({
  Card: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  CardContent: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  CardFooter: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  CardHeader: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  CardTitle: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
}));
jest.mock("@/components/ui/skeleton", () => ({
  Skeleton: () => <div data-testid="skeleton" />,
}));
jest.mock("lucide-react", () => ({
  ArrowRight: () => <span data-testid="arrow-right" />,
}));
jest.mock("next/link", () => ({
  __esModule: true,
  default: (props: Record<string, unknown>) => <a {...props} />,
}));

describe("DepositPage", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    getOffers.mockResolvedValue({
      data: [
        {
          id: 1,
          merchant_display_name: "Merchant 1",
          price: 1,
          min_amount: 100000,
          max_amount: 10000000,
          payment_details: JSON.stringify({
            bank_name: "Bank",
            account_number: "123456",
            account_holder_name: "Test User",
          }),
          currency: "vnd",
          country_code: "VN",
          offer_type: "sell",
          is_active: true,
          online: true,
          user_id: 2,
        },
      ],
    });
    createTrade.mockResolvedValue({ id: 123 });
  });

  it("should render loading skeleton and content", () => {
    const { getByTestId } = render(<DepositPage />);
    expect(getByTestId("protected-layout")).toBeInTheDocument();
    expect(getByTestId("fiat-transaction-layout")).toBeInTheDocument();
  });

  it("should show offers after entering valid amount and clicking continue", async () => {
    const { getByLabelText, getByText, findByText } = render(<DepositPage />);
    const input = getByLabelText("amountToDeposit");
    fireEvent.change(input, { target: { value: "200000" } });
    fireEvent.click(getByText("continue"));
    await findByText("Merchant 1");
    expect(getOffers).toHaveBeenCalled();
  });

  it("should show error toast if no offers available", async () => {
    getOffers.mockResolvedValue({ data: [] });
    const { getByLabelText, getByText, findByText } = render(<DepositPage />);
    const input = getByLabelText("amountToDeposit");
    fireEvent.change(input, { target: { value: "200000" } });
    fireEvent.click(getByText("continue"));
    await waitFor(() => {
      expect(toast).toHaveBeenCalledWith(
        expect.objectContaining({ title: "noOffersForCurrency" }),
      );
    });
    await findByText("noMerchantAvailable");
  });

  it("should call createTrade and redirect on confirm deposit", async () => {
    const { getByLabelText, getByText, findByText } = render(<DepositPage />);
    const input = getByLabelText("amountToDeposit");
    fireEvent.change(input, { target: { value: "200000" } });
    fireEvent.click(getByText("continue"));
    await findByText("Merchant 1");
    fireEvent.click(getByText("Merchant 1"));
    fireEvent.click(getByText("confirmDeposit"));
    await waitFor(() => {
      expect(createTrade).toHaveBeenCalled();
      expect(push).toHaveBeenCalledWith("/trade/123");
    });
  });

  it("should show error toast if createTrade fails", async () => {
    createTrade.mockRejectedValueOnce(new Error("fail"));
    const { getByLabelText, getByText, findByText } = render(<DepositPage />);
    const input = getByLabelText("amountToDeposit");
    fireEvent.change(input, { target: { value: "200000" } });
    fireEvent.click(getByText("continue"));
    await findByText("Merchant 1");
    fireEvent.click(getByText("Merchant 1"));
    fireEvent.click(getByText("confirmDeposit"));
    await waitFor(() => {
      expect(toast).toHaveBeenCalledWith(
        expect.objectContaining({ title: "errorCreatingDeposit" }),
      );
    });
  });

  it("should match snapshot", () => {
    const { container } = render(<DepositPage />);
    expect(container).toMatchSnapshot();
  });

  it("should not show offers when amount is empty", async () => {
    const { getByText } = render(<DepositPage />);
    fireEvent.click(getByText("continue"));
    expect(getOffers).not.toHaveBeenCalled();
  });

  it("should show error toast when amount is less than minimum", async () => {
    const { getByLabelText, getByText } = render(<DepositPage />);
    const input = getByLabelText("amountToDeposit");
    fireEvent.change(input, { target: { value: "10000" } }); // Less than 50,000
    fireEvent.click(getByText("continue"));
    await waitFor(() => {
      expect(toast).toHaveBeenCalledWith(
        expect.objectContaining({
          title: "invalidAmount",
          description: "amountTooLow",
          variant: "destructive",
        }),
      );
    });
    expect(getOffers).not.toHaveBeenCalled();
  });

  it("should show error toast if getOffers throws", async () => {
    getOffers.mockRejectedValueOnce(new Error("fail"));
    const { getByLabelText, getByText } = render(<DepositPage />);
    const input = getByLabelText("amountToDeposit");
    fireEvent.change(input, { target: { value: "200000" } });
    fireEvent.click(getByText("continue"));
    await waitFor(() => {
      expect(toast).toHaveBeenCalledWith(
        expect.objectContaining({ title: "errorFetchingOffers" }),
      );
    });
  });

  it("should show error toast if selected offer not found", async () => {
    const { getByLabelText, getByText, findByText } = render(<DepositPage />);
    const input = getByLabelText("amountToDeposit");
    fireEvent.change(input, { target: { value: "200000" } });
    fireEvent.click(getByText("continue"));
    await findByText("Merchant 1");
    // Chọn offerId không tồn tại
    fireEvent.click(getByText("Merchant 1"));
    // Xóa hết offers trong state bằng cách set offers rỗng qua mock
    // Giả lập selectedOfferId vẫn còn nhưng offers rỗng
    // Cần mock lại useState để trả về [] cho offers
    // Đơn giản nhất: mock createTrade throw error như case fail
    createTrade.mockImplementationOnce(() => {
      throw new Error("Selected offer not found");
    });
    fireEvent.click(getByText("confirmDeposit"));
    await waitFor(() => {
      expect(toast).toHaveBeenCalledWith(
        expect.objectContaining({ title: "errorCreatingDeposit" }),
      );
    });
  });

  it("should go back to amount input when clicking back button if no offers", async () => {
    getOffers.mockResolvedValue({ data: [] });
    const { getByLabelText, getByText, findByText } = render(<DepositPage />);
    const input = getByLabelText("amountToDeposit");
    fireEvent.change(input, { target: { value: "200000" } });
    fireEvent.click(getByText("continue"));
    await findByText("noMerchantAvailable");
    fireEvent.click(getByText("back"));
    // Sau khi back, nút continue lại xuất hiện
    expect(getByText("continue")).toBeInTheDocument();
  });
});
