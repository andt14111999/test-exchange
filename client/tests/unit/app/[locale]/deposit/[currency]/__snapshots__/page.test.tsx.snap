// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DepositPage should match snapshot 1`] = `
<div>
  <div
    data-testid="protected-layout"
  >
    <div
      data-testid="fiat-transaction-layout"
    >
      <div
        class="space-y-8"
      >
        <div
          class="space-y-6"
        >
          <div
            class="space-y-4"
          >
            <div
              class="flex justify-end"
            >
              <p
                class="text-sm text-muted-foreground"
              >
                balance
                :
                 
                1,000,000
                 
                VND
              </p>
            </div>
            <div
              class="space-y-2"
            >
              <label
                for="amount"
              >
                amountToDeposit
              </label>
              <input
                class="pl-2"
                id="amount"
                inputmode="numeric"
                placeholder="50,000 VND"
                type="text"
                value=""
              />
              <p
                class="text-sm text-muted-foreground"
              >
                min
                :
                 
                50,000
                 
                VND
              </p>
            </div>
            <div
              data-testid="skeleton"
            />
          </div>
          <button
            class="w-full"
            disabled=""
          >
            continue
          </button>
        </div>
        <div>
          <div>
            <div>
              depositHistory
            </div>
          </div>
          <div>
            <p
              class="text-sm text-muted-foreground"
            >
              viewHistoryDescription
            </p>
          </div>
          <div>
            <button
              variant="outline"
            >
              <a
                href="/wallet/history/fiat/vnd"
              >
                viewHistory
                <span
                  data-testid="arrow-right"
                />
              </a>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
